### PR TITLE
[6.x/7.x] Scenario System

### LINKED ISSUE #17905: [6.x/7.x] Scenario System (50$ bounty)
**Description**:
Read this: http://wow.gamepedia.com/Scenario

> From Warlords of Draenor onwards, there are no new endgame scenarios, and thus, no scenario rewards. Instead, some quests use scenario mechanics, which are seamlessly integrated into the main world using phasing, rather than being their own instances.

Scenarios are like additional Quests without typical Questgiver Questender Mechanics.
It's to provide content for players. Scenarios are most time used in OpenWorld as some story content.
In Dungeons it shall guide through the dungeon and reward people if completing it.

**Current behaviour**: No Implementation 

**Expected behaviour**: 
- Dungeon scenarios are stored inside DB2 file
- World must be stored inside WorldDB or HotfixesDB or it's via AchievementCriteria.
- State of scenario must be saved and loaded
- Packets must be updated and researched (Unk Bits)
- Entering a WorldScenario with POI must be displayed on Map
- Achievements related to Scenario must be granted
- Bonus Objectives must be loaded and rewarded

Is there a scaling with itemlevel?
for example the artifact scenarios.

**Branch(es)**: 6.x

**TDB version**:  6.04

**Operating system**:  Win 10

**Bountysource**
https://www.bountysource.com/issues/37599527-6-x-7-x-scenario-system

[Bountysource](https://www.bountysource.com/issues/37599527-6-x-7-x-scenario-system-50-bounty?utm_source=37599527&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**forked Branches**
https://github.com/Krudor/TrinityCore/tree/scenarios_cleanup

### PR DESCRIPTION
**Pull request by Krudor**
**Changes proposed:**
- Add base Scenario system template  
- Implement Instance Scenarios

**Features**
- [x] Instance Scenario progress is saved when the instance is saved
- [x] Scenario linked to instances through db table 'world.scenarios'
- [x] Scenario POI's fully functional (Map&Minimap displays for scenario objectives)
- [x] Scenario Step Progress
- [x] Scenario Bonus Objectives
- [x] Scenario Step / Bonus Objectives Rewards

**Target branch(es):** 6.x

**Issues addressed:** Closes #17905

**Tests performed:** In game tested

**Known issues and TODO list:**
- Zone Scenarios (Not planned for implementation in current PR)
- Other types of scenarios (Not planned for implementation in current PR)

There seems to be an issue related to reusing old instance ids.
Progressing through an instance, saving, and unbinding the instance from a player, and re-entering the instance seems to sometimes reuse old instance ids.
This causes a problem because instance scenario progress is saved for each instance id, and reusing old ones comes with loading old instance scenario progress data.

**General TODO list for Scenarios in general:**
- Figure out unknown SMSG_SCENARIO_STATE opcode data
- Figure out unknown SMSG_SCENARIO_BOOT opcode data
- Implement more varieties of scenarios such as zone scenarios by using Scenario as a base class
- Add data in world.scenarios table to link scenarios to the new Legion instances, so far only WoD and < is linked.
- Send scenario event updates to InstanceScript hooks?

**Overview**
Implements instance scenarios.
Any dungeons properly linked in the database 'world.scenarios' will properly display the scenario interface when entering an instance.
The scenario system can be expanded upon, for example we don't send any events that scripts can hook into whenever a step criteria changes, which is required to script some scenarios, but I think it can wait for now.

### CODE REVIEW

File: src/server/game/Server/Packets/ScenarioPackets.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+                WorldPacket const* Write() override;
+
+                WorldPackets::Achievement::CriteriaProgress CriteriaProgress;
+        };
+
+        class ScenarioCompleted final : public ServerPacket
+        {
+            public:
+                ScenarioCompleted(uint32 scenarioId) : ScenarioId(scenarioId), ServerPacket(SMSG_SCENARIO_COMPLETED, 4) { }
```
Comment (tkrokli) [inline src/server/game/Server/Packets/ScenarioPackets.h]:
```
In file included from /home/travis/build/TrinityCore/TrinityCore/src/server/game/PrecompiledHeaders/gamePCH.cpp:1:
In file included from /home/travis/build/TrinityCore/TrinityCore/src/server/game/PrecompiledHeaders/gamePCH.h:11:
In file included from /home/travis/build/TrinityCore/TrinityCore/src/server/game/Server/Packets/AllPackets.h:63:
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Server/Packets/ScenarioPackets.h:74:56: fatal error: 
      field 'ScenarioId' will be initialized after base
      'WorldPackets::ServerPacket' [-Wreorder]
                ScenarioCompleted(uint32 scenarioId) : ScenarioId(scenar...
                                                       ^
```

File: src/server/game/Scenarios/Scenario.cpp [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        criteriaProgress.Flags = timedCompleted ? 1 : 0;
+
+    criteriaProgress.TimeFromStart = timeElapsed;
+    criteriaProgress.TimeFromCreate = 0;
+
+    progressUpdate.CriteriaProgress = criteriaProgress;
+    SendPacket(progressUpdate.Write());
+}
+
+void Scenario::SendCriteriaProgressRemoved(uint32 criteriaId)
```
Comment (tkrokli) [inline src/server/game/Scenarios/Scenario.cpp]:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Scenarios/Scenario.cpp:147:51: fatal error: 
      unused parameter 'criteriaId' [-Wunused-parameter]
void Scenario::SendCriteriaProgressRemoved(uint32 criteriaId)
                                                  ^
```

File: src/server/game/Scenarios/Scenario.cpp [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+    if (step->IsBonusObjective())
+        return !IsComplete();
+
+    if (step != GetStep())
+        return false;
+
+    return true;
+}
+
+void Scenario::CompletedCriteriaTree(CriteriaTree const * tree, Player * referencePlayer)
```
Comment (tkrokli) [inline src/server/game/Scenarios/Scenario.cpp]: Thank you for the previous follow-ups, but there is still more to modify:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Scenarios/Scenario.cpp:184:74: fatal error: 
      unused parameter 'referencePlayer' [-Wunused-parameter]
  ...const * tree, Player * referencePlayer)
                            ^
```

Comment (Krudor) [inline src/server/game/Scenarios/Scenario.cpp]: Thanks tkrokli, I'll see if I can find more unused parameters other than that for the next commit :)

Comment (tkrokli) [inline src/server/game/Scenarios/Scenario.cpp]: Good job, looks like you nailed it this time. 👍

**Comment by Shauren** [comment]
Please use https://github.com/Shauren/WowClientDB2MySQLTableGenerator to regenerate HotfixDatabase files and DB2LoadInfo.h header

**Comment by Krudor** [comment]
Crap, I did not read the header of DB2LoadInfo.h, apologies!

**Comment by Palabola** [comment]
Any estimated time about merging it to master?

**Comment by Aokromes** [comment]
No.

**Comment by Krudor** [comment]
I'm a bit busy right now, but I'm going to update some autogenerated files now that Shauren fixed a bug in the generator that generated those files.

**Comment by Aokromes** [comment]
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.
Conflicting files
sql/base/characters_database.sql

**Comment by funjoker** [comment]
Did anyone test it ?

**Comment by tkrokli** [comment]
@funjoker : I don't think so, only Krudor. Otherwise you would have seen more comments, I think.

**Pull request closed without merging**